aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -103,18 +108,83 @@ compute_result_type(at::ArrayRef<OperandInfo> operands,
 
 std::tuple<Device, ScalarType> TensorIterator::compute_common_type() {
   // See [Result type computation] in TensorIterator.h
+
   auto result_type =
       compute_result_type(operands_,
-        [](const Tensor& t) { return t.dim() > 0; },
-        [](const Tensor& t) { return !t.unsafeGetTensorImpl()->is_wrapped_number(); },
-        [](const Tensor& t) { return true; });
+        [](const OperandInfo& op) { return op.tensor.dim() > 0; },
+        [](const OperandInfo& op) { return !op.tensor.unsafeGetTensorImpl()->is_wrapped_number(); },
+        [](const OperandInfo& op) { return true; });
+
+  if (ScalarType::Bool == std::get<1>(result_type)) {
+    auto alternate = compute_result_type(operands_,
+        [](const OperandInfo& op) {
+          return op.tensor.dim() == 0;
+        }
+    );
+    if (std::get<1>(alternate) != ScalarType::Undefined) {
+      // preserve device from original result
+      return std::make_tuple(std::get<0>(result_type), std::get<1>(alternate));
+    }
+  }
+
+  // if non-zero-dim tensor result is an integral type and there's a zero-dim
+  // floating point operand, we'll promote the floating point type.
+  if (isIntegralType(std::get<1>(result_type), false)) {
+    auto alternate = compute_result_type(operands_,
+        [](const OperandInfo& op) {
+          return isFloatingType(op.tensor.scalar_type()) && op.tensor.dim() == 0;
+        }
+    );
+    if (std::get<1>(alternate) != ScalarType::Undefined) {
+      // preserve device from original result
+      return std::make_tuple(std::get<0>(result_type), std::get<1>(alternate));
+    }
+  }
 
   TORCH_INTERNAL_ASSERT(std::get<1>(result_type) != ScalarType::Undefined);
   return result_type;
 }
 
+static void validate_dtype(OperandInfo& op, ScalarType common_dtype, int ninputs) {
+  if (op.tensor.defined()) {
+    // For binary_ops, we follow casting rules. For unary/nullary types
```

Comment:
what about > 2 operands?

Reply:
We currently just have binary, unary, nullary, and reduce ops. extra scalar operands like `alpha` don't count